base/dict.jl

```diff
@@ -258,11 +258,12 @@ function empty!(h::Dict{K,V}) where V where K
 end
 
 # get the index where a key is stored, or -1 if not present
-function ht_keyindex(h::Dict{K,V}, key) where V where K
+@assume_effects :terminates_locally function ht_keyindex(h::Dict{K,V}, key) where V where K
     isempty(h) && return -1
     sz = length(h.keys)
     iter = 0
     maxprobe = h.maxprobe
+    @assert maxprobe <= sz # needed for terminates_locally to be valid
```

Comment:
Is `@assert` the correct way of checking this? It doesn't guarantee the condition to be true after all, since it may be removed by the compiler (although I don't think it currently is).

Reply:
good point.